src/validation.cpp

```diff
@@ -4959,14 +4965,17 @@ void ChainstateManager::CheckBlockIndex()
             assert((pindex->nStatus & BLOCK_FAILED_MASK) == 0); // The failed mask cannot be set for blocks without invalid parents.
         }
         // Make sure nChainTx sum is correctly computed.
-        unsigned int prev_chain_tx = pindex->pprev ? pindex->pprev->nChainTx : 0;
-        assert((pindex->nChainTx == pindex->nTx + prev_chain_tx)
-               // Transaction may be completely unset - happens if only the header was accepted but the block hasn't been processed.
-               || (pindex->nChainTx == 0 && pindex->nTx == 0)
-               // nChainTx may be unset, but nTx set (if a block has been accepted, but one of its predecessors hasn't been processed yet)
-               || (pindex->nChainTx == 0 && prev_chain_tx == 0 && pindex->pprev)
-               // Transaction counts prior to snapshot are unknown.
-               || pindex->IsAssumedValid());
+        if (!pindex->pprev) {
+            // If no previous block, nTx and nChainTx must be the same.
+            assert(pindex->nChainTx == pindex->nTx);
+        } else if (pindex->pprev->nChainTx > 0 && pindex->nTx > 0) {
+            // If previous nChainTx is set and number of transactions in block is known, sum must be set.
+            assert(pindex->nChainTx == pindex->nTx + pindex->pprev->nChainTx);
+        } else {
+            // Otherwise nChainTx should only be set if this is a snapshot block.
+            assert((pindex->nChainTx != 0) == (pindex == snap_base));
```

Comment:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1509611419

> This is also violated in the situation of a snapshot block that is sent out of order (`pindex->nChainTx == 0`, `pindex == snap_base` )

Ok, I guess it needs to be changed to:

```c++
assert(pindex->nChainTx != 0 || pindex == snap_base);
```

Reply:
re: https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1509611419

Could reproduce this failure with the test change suggested in https://github.com/bitcoin/bitcoin/pull/29370#issuecomment-1974096225. It actually needed a slightly different fix `assert(pindex->nChainTx == 0 || pindex == snap_base || pindex->pprev == snap_base)` because there if the snapshot block temporarily had nChainTx set to 0, and the block after the snapshot block was previously downloaded, the assert could trigger again without the `pindex->pprev == snap_base` condition. Explaining this in another code comment seemed too complicated though, so opted to revert this and make a fix in `ReceivedBlockTransactions` instead